### ISSUE TITLE
When using EventAwareLumiBased splitting, some jobs analyze less lumis than expected

### ISSUE THREAD

**Issue by AndresTanasijczuk** [dmwm/WMCore#5907]
I am testing CRAB3 with EventAwareLumiBased splitting (I am adding the feature of allowing to specify a "total_events" limit to stop creating new jobs; similar as I did for LumiBased and FileBased splitting in the past) and I observed that some jobs analyze less lumis than expected, in the sense that some jobs could analyze more lumis than they did and still fit within the requirement to analyze (approximately and at most) N (expected) events. I have identified the problem as coming from the fact that, when changing input files, the counting of how many events are already included in the job in line
https://github.com/dmwm/WMCore/blob/1.0.7.pre3/src/python/WMCore/JobSplitting/EventAwareLumiBased.py#L291
may use f['lumiCount'], which includes all lumis in the file, not only the good lumis filtered by the lumi mask. There is a min(lumisInJob, f['lumiCount']) in the counting, so when lumisInJob <= f['lumiCount'] there is not problem, since lumisInJob does count only good lumis, but when f['lumiCount'] < lumisInJob and therefore f['lumiCount'] is used in the counting, this includes all lumis in the file (https://github.com/dmwm/WMCore/blob/1.0.7.pre3/src/python/WMCore/JobSplitting/EventAwareLumiBased.py#L112-119), not only the good ones.

For my own information, the task where I observed the problem is 150513_154848:atanasi_crab_test_totalUnits_lumimask_7K_30K_without_final_fix (http://glidemon.web.cern.ch/glidemon/jobs.php?taskid=1093404) in particular job 2.

**Comment by PerilousApricot** [comment]
I'm confused, shouldn't you be limiting "total_lumis"? That splitter is
lumi based splitting, it just takes into account some lumis being bigger
than others

-- 
Sent from Gmail Mobile

**Comment by ericvaandering** [comment]
It splits by lumis but uses events as the suggested unit. Obviously it's an approximation.

When I discussed this on the tram with @mmascher I suggested that the "right" thing to do is to make each job no larger than the suggested units_per_job (in events and if possible, of course) but make sure the task is at least as large as "total_units". So cut off after the lumi that gets them *over* total_units.
